pkg/iter/entry_iterator.go

```diff
@@ -347,36 +376,58 @@ func (i *entrySortIterator) init() {
 		}
 		util.LogError("closing iterator", it.Close)
 	}
-	heap.Init(i.heap)
+	i.is = tmp
+	sort.Slice(i.is, i.lessByIndex)
+}
 
-	// We can now clear the list of input iterators to merge, given they have all
-	// been processed and the non empty ones have been pushed to the heap
-	i.is = nil
+func (i *entrySortIterator) fix() {
+	head := i.is[0]
+	t1 := head.Entry().Timestamp.UnixNano()
+	l1 := head.Labels()
+	h1 := head.StreamHash()
+
+	// shortcut
+	if len(i.is) <= 1 || i.lessByValue(t1, l1, h1, 1) {
+		return
+	}
+
+	// First element is out of place. So we reposition it.
+	i.is = i.is[1:] // drop head
```

Comment:
I'm sure we can factor this into line 401 but it gave me too much of a headache.